frontend/src/app/main/ui/ds/utilities/swatch.cljs

```diff
@@ -0,0 +1,101 @@
+;; License, v. 2.0. If a copy of the MPL was not distributed with this
+;; file, You can obtain one at http://mozilla.org/MPL/2.0/.
+;;
+;; Copyright (c) KALEIDOS INC
+
+(ns app.main.ui.ds.utilities.swatch
+  (:require-macros
+   [app.main.style :as stl])
+  (:require
+   [app.common.data.macros :as dm]
+   [cuerdas.core :as str]
+   [rumext.v2 :as mf]))
+
+(def ^:private schema:swatch
+  [:map
+   [:background :string]
+   [:aria-label {:optional true} :string]
+   [:format {:optional true} [:enum "square" "rounded"]]
+   [:size {:optional true} [:enum "small" "medium"]]
+   [:active {:optional true} :boolean]
+   [:on-click {:optional true} fn?]])
+
+(def hsla-regex #"^hsla\((\d{1,3}),\s*(\d{1,3})%,\s*(\d{1,3})%,\s*(0|1|0?\.\d+)\)$")
+(def rgba-regex #"^rgba\((\d{1,3}),\s*(\d{1,3}),\s*(\d{1,3}),\s*(0|1|0?\.\d+)\)$")
+
+(defn- is-image? [background]
+  (boolean
+   (str/starts-with? background "http")))
+
+(defn- is-gradient? [background]
+  (boolean
+   (or
+    (str/starts-with? background "linear-gradient")
+    (str/starts-with? background "radial-gradient"))))
+
+(defn- is-color-opacity? [background]
+  (boolean
+   (or (re-matches hsla-regex background)
+       (re-matches rgba-regex background))))
+
+(defn- extract-color-and-opacity [background]
+  (cond
+    (re-matches rgba-regex background)
+    (let [[_ r g b a] (re-matches rgba-regex background)]
+      {:color (dm/str "rgb(" r ", " g ", " b ")")
+       :opacity (js/parseFloat a)})
+
+    (re-matches hsla-regex background)
+    (let [[_ h s l a] (re-matches hsla-regex background)]
+      {:color (dm/str "hsl(" h ", " s "%, " l "%)")
+       :opacity (js/parseFloat a)})
+
+    :else
+    {:color background
+     :opacity 1.0}))
+
+(mf/defc swatch*
+  {::mf/props :obj
+   ::mf/schema schema:swatch}
+  [{:keys [background on-click format size aria-label active] :as props}]
+  (let [element-type (if on-click "button" "div")
+        format (or format "square")
+        size (or size "small")
+        active (or active false)
+        {:keys [color opacity]} (extract-color-and-opacity background)
+        on-click (mf/use-fn (mf/deps on-click)
+                            (fn [event]
+                              (when (fn? on-click)
+                                (on-click event))))]
+    [:> element-type {:class (stl/css-case
+                              :swatch true
+                              :small (= size "small")
+                              :medium (= size "medium")
+                              :square (= format "square")
+                              :active (= active true)
+                              :interactive (= element-type "button")
+                              :rounded (= format "rounded"))
+                      :aria-label (when on-click aria-label)
+                      :on-click (when on-click on-click)}
+
+     (cond
+       (is-image? background)
+       [:div {:class (stl/css :swatch-image)
```

Comment:
shouldn't all this `<div>` be `<button>` elements?

Reply:
No, there is only a common button that wraps all these elements when the swatch is interactive. These are the specific implementation of the swatch background type.

Reply:
But why have an external wrapper when we can do it here? I see two alternatives:

- Always use `<button>` and set `disable` when not interactive.
- Render either a `<button>` or a `<div>` depending if it's interactive or not.

Accessibility-wise, if something is interactive it should have the proper aria semantics. `<button>` give us those for free (and also sets activation via keyboard, etc. for free)